src/acc/fasten.hpp

```diff
@@ -158,78 +158,77 @@ template <size_t PPWI> class IMPL_CLS final : public Bude<PPWI> {
+      protein[:natpro], ligand[:natlig],      \
+      forcefield[:ntypes], poses_0[:nposes],  \
+      poses_1[:nposes], poses_2[:nposes],     \
+      poses_3[:nposes], poses_4[:nposes],     \
+      poses_5[:nposes])                       \
+    create(energies[:nposes])
+
+    auto hostToDeviceEnd = now();
+    sample.hostToDevice = {hostToDeviceStart, hostToDeviceEnd};
+
+    // clang-format on
+    for (size_t i = 0; i < p.totalIterations(); ++i) {
+      auto kernelStart = now();
+
+    // clang-format off
+  #pragma acc parallel loop                   \
+    present(                                  \
+      protein[:natpro], ligand[:natlig],      \
+      forcefield[:ntypes], poses_0[:nposes],  \
+      poses_1[:nposes], poses_2[:nposes],     \
+      poses_3[:nposes], poses_4[:nposes],     \
+      poses_5[:nposes], energies[:nposes])
       // clang-format on
-      for (size_t i = 0; i < p.totalIterations(); ++i) {
-        auto kernelStart = now();
-
-// clang-format off
-#pragma acc parallel loop  \
-        present(                                    \
-            protein[:natpro], ligand[:natlig],      \
-            forcefield[:ntypes], poses_0[:nposes],  \
-            poses_1[:nposes], poses_2[:nposes],     \
-            poses_3[:nposes], poses_4[:nposes],     \
-            poses_5[:nposes], energies[:nposes]) wait
-        // clang-format on
-        for (size_t group = 0; group < (nposes / PPWI); group++) {
-          fasten_main<PPWI>(group, ntypes, nposes, natlig, natpro,                //
-                            protein, ligand,                                      //
-                            poses_0, poses_1, poses_2, poses_3, poses_4, poses_5, //
-                            forcefield, energies);
-        }
-        auto kernelEnd = now();
-        sample.kernelTimes.emplace_back(kernelStart, kernelEnd);
+      for (size_t group = 0; group < (nposes / PPWI); group++) {
+        fasten_main<PPWI>(group, ntypes, nposes, natlig, natpro,                //
+                          protein, ligand,                                      //
+                          poses_0, poses_1, poses_2, poses_3, poses_4, poses_5, //
+                          forcefield, energies);
       }
+      auto kernelEnd = now();
+      sample.kernelTimes.emplace_back(kernelStart, kernelEnd);
     }
+
+  auto deviceToHostStart = now();
+        
+  #pragma acc update                          \
+    host(energies[:nposes])
+        
+  auto deviceToHostEnd = now();
+  sample.deviceToHost = {deviceToHostStart, deviceToHostEnd};
+  
+  #pragma acc exit data                       \
+    delete(protein, ligand, forcefield,       \
+           poses_0, poses_1, poses_2,         \
+           poses_3, poses_4, poses_5,         \
+           energies)
+
     std::copy(energies, energies + p.nposes(), sample.energies.begin());
-    std::free(protein);
-    std::free(ligand);
-    std::free(forcefield);
-    std::free(energies);
```

Comment:
`energies` is calloc'd above so I believe we need to free it before the name goes out of scope.

Reply:
Or change energies to point to sample.energies